src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,211 @@
+        SampleUtil::copy(pBuffer, m_buffer.data(), itemsToRead - firstDataBlockSize);
+    } else {
+        // Read is contiguous.
+        SampleUtil::copy(pBuffer, m_buffer.data(m_readPos), itemsToRead);
+    }
+
+    // Calculate the new read position. If the new read position
+    // is after the ring delay buffer end, move it around from the start
+    // of the ring delay buffer.
+    m_readPos = (m_readPos + itemsToRead);
+
+    if (m_readPos >= m_buffer.size()) {
+        m_readPos = m_readPos - m_buffer.size();
+        // The read position crossed the right bound of the buffer. So, if the situation,
+        // where the read position equals the write position arises, and before this situation,
+        // the write position does not cross the right bound of the buffer too,
+        // the buffer will be empty.
+        m_fullFlag = false;
+    }
+
+    return itemsToRead;
+}
+
+SINT RingDelayBuffer::write(const CSAMPLE* pBuffer, const SINT numItems) {
+    const SINT available = getWriteAvailable();
+    SINT itemsToWrite = numItems;
+
+    VERIFY_OR_DEBUG_ASSERT(itemsToWrite <= available) {
+        itemsToWrite = available;
+    }
+
+    if (m_firstInputBuffer) {
+        // If the first input buffer is written, the first sample is on the index 0.
+        // Based on the checking of an available number of samples, the situation,
+        // that the writing will be non-contiguous cannot occur.
+        // The itemsToWrite value is multiply by 2 to
+        SampleUtil::copyWithRampingGain(m_buffer.data(), pBuffer, 0.0f, 1.0f, itemsToWrite);
+
+        m_firstInputBuffer = false;
+    } else if ((m_writePos + itemsToWrite) > m_buffer.size()) { // Check to see
+                                                                // if the write is not contiguous.
+        // Write is not contiguous.
+        SINT firstDataBlockSize = m_buffer.size() - m_writePos;
+
+        SampleUtil::copy(m_buffer.data(m_writePos), pBuffer, firstDataBlockSize);
+        pBuffer = pBuffer + firstDataBlockSize;
+
+        // The second data part is the start of the ring buffer.
+        SampleUtil::copy(m_buffer.data(), pBuffer, itemsToWrite - firstDataBlockSize);
+    } else {
+        // Write is contiguous.
+        SampleUtil::copy(m_buffer.data(m_writePos), pBuffer, itemsToWrite);
+    }
+
+    // Calculate the new write position. If the new write position
+    // is after the ring delay buffer end, move it around from the start
+    // of the ring delay buffer.
+    m_writePos = (m_writePos + itemsToWrite);
+
+    if (m_writePos >= m_buffer.size()) {
+        m_writePos = m_writePos - m_buffer.size();
+        // The write position crossed the right bound of the buffer. So, if the situation,
+        // where the write position equals the read position arises, and before this situation,
+        // the read position does not cross the right bound of the buffer too,
+        // the buffer will be full.
+        m_fullFlag = true;
```

Comment:
The buffer is full if itemsToWrite >= available, not due to crossing the boundaries.

Reply:
Right, thank you for this point, it really didn't make much sense and was hard to understand the flag handling. The new version should fix this based on the proposed change.